File: python3/palindromic_substrings.py
"""
Space   : O(1)
Time    : O(n)

Method:
Palindrome stack
"""


class Solution:
    def countSubstrings(self, s: str) -> int:
        ans = 0
        n = len(s)
        i = 0

        while i < n:
            start, end = i, i
            ans += 1

            # expand forward if next char is equal
            while (end + 1 < n) and (s[end] == s[end + 1]):
                end += 1
                ans += 1

            # expand start and end if char at start and end are equal
            while (end + 1 < n) and (start - 1 >= 0) and (s[start-1] == s[end+1]):
                end += 1
                start -= 1
                ans += 1

            i += 1

        return ans

File: golang/palindromic_substrings.go
package main

/*
Space	: O(1)
Time	: O(n)

Method:
Palindrome stack
*/


func countSubstrings(s string) int {
	ans, n, i := 0, len(s), 0

	for i < n {
		start, end := i, i
		ans++

		for (end+1 < n) && (s[end] == s[end+1]) {
			end++
			ans++
		}

		for (end+1 < n) && (start-1 >= 0) && (s[start-1] == s[end+1]) {
			start--
			end++
			ans++
		}

		i++
	}

	return ans
}
